# TECT

Please aligned by cellranger. It will give the gene expression. Otherwise, it is difficult to deal with different gene expression matrices in a unified format.

## Installation

It is recommended to use conda to create a virtual environment in python 3.7.9. Then run the following code to install the required package.

```
$ git clone https://github.com/yzf072/TECT
$ cd TECT
$ pip install -r requirements.txt
$ python setup.py install
```

## Index part

Construct the TE Library index based on the interval Tree

### input

* TE annotation file(include the **genomic location of TEs**)

  Here, we use repeatmasker genomic datasets. It is generated by analyzing the complete genomes with RepeatMasker, a program that screens DNA sequences for interspersed repeats,  and offer the results in various formats. These datasets were generated using the computing resources at UCSC. 

  Let's show some lines. It contains the information of **repeat type(include repeat class/family and specific individual)** and **starting position** in different position of each chromosome. 

  ```
     SW  perc perc perc  query      position in query           matching       repeat              position in  repeat
  score  div. del. ins.  sequence    begin     end    (left)    repeat         class/family         begin  end (left)   ID
  
    535  21.2 15.9  3.1  chr1        11485   11676 (248944746) C  L1MC5a         LINE/L1              (510) 5667   5447      3
  
  ```

  Here is the web address. 

  Human:

  http://www.repeatmasker.org/genomes/hg38/RepeatMasker-rm405-db20140131/hg38.fa.out.gz

  Mouse:

  http://www.repeatmasker.org/genomes/mm10/RepeatMasker-rm405-db20140131/mm10.fa.out.gz

  And we can choose other species in this page, please download the newest vesion file in 'xxx.fa.out.gz' format.

  https://www.repeatmasker.org/genomicDatasets/RMGenomicDatasets.html

* gene annotation file(include genome infromation)

  We use it to count how many reads are aligned to the exon region. 

  Its content is like this.

  ```
  chr1	HAVANA	gene	11869	14409	.	+	.	ID=ENSG00000223972.5;gene_id=ENSG00000223972.5;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;level=2;hgnc_id=HGNC:37102;havana_gene=OTTHUMG00000000961.2
  chr1	HAVANA	transcript	11869	14409	.	+	.	ID=ENST00000456328.2;Parent=ENSG00000223972.5;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1
  chr1	HAVANA	exon	11869	12227	.	+	.	ID=exon:ENST00000456328.2:1;Parent=ENST00000456328.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;exon_number=1;exon_id=ENSE00002234944.1;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1
  ```

   It contains the comprehensive gene annotation on the reference chromosomes only. Through this, we could get the count of the reads on exon region. Then we use the count to normalize the count of TE.

  Human:

  http://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens

  Mouse:

  http://ftp.ensembl.org/pub/release-104/gtf/mus_musculus

  You could get other species' gtf file on Ensemble:

  http://asia.ensembl.org/info/data/ftp/index.html

### Usage

Prepare the TE and gene annotation file to build the TE library index.

```shell
$ TECT_build -a te_annotation.out -g exon.gff3 -o human.index/
```

More detailed help

```shell
$ TECT_build -h
usage: TECT_build [-h] -a ANNOTATION -g GFF [-o OUTPUT]

Construct the TE Library index based on the interval Tree.

optional arguments:
  -h, --help            show this help message and exit

required arguments:
  -a ANNOTATION, --annotation ANNOTATION
                        Give the path of TE annotation file.
  -g GFF, --gff GFF     Give the path of gff file. Help to determine whether a gene is on the exon or not.
  -o OUTPUT, --output OUTPUT
                        Prefix path of the output file.

Example: TECT_build -a te_annotation.out -g exon.gff3 -o human.index/
```

## Count part

Get the expression of TE from single cell RNA sequencing data.

### input

* aligned sequence reads

  You will need to use the alignment tool to map the reads to the reference genome to get the bam format file including each read's location. STAR is recommended.

  Our pipeline can deal with **C1/SMART-seq-style dat**a, each cell corresponds to a fastq file, and the **10x style data**, reads from different cells are distincted by the CB/CR tag. 

  

  **!attention! If the data is in 10x style, you will have to use samtool to sort bam files by barcode.** 

  If data was aligned by cellranger, the barcode will be stored in CB tag. And if data was aligned by STAR, the barcode will be stored in CR tag.

  Cellranger:

  ```shell
  $ samtools sort -t CB -o bcsorted.bam xxx.bam
  ```

  STAR:

  ```shell
  $ samtools sort -t CR -o bcsorted.bam xxx.bam
  ```

* index file

  Give the path of the index file you generated on the TECT_build step.

### Usage

Give the bam file and index file, run the TECT module, then wait for the cell x TE matrix.

```shell
$ TECT -i xxx.bam -idx human.index/index.idx -CB None -o result
```

* smart-seq style data

  Set -CB None, the default setting is.

* 10x style data

  Aligned by **cellranger**, set -CB CB.

  Aligned by **STAR**, set -CB CR.

If there are more than one bam file, you can separate them with the comma, or *.bam:

```shell
$ TECT -i xxx.bam,yyy.bam,zzz.bam -idx human.index/index.idx -CB None -o result
$ TECT -i *.bam -idx human.index/index.idx -CB CB -o result
```

There are some optional arguments you could adjust.

```shell
$ TECT -h                                                                                                                                                         
usage: TECT [-h] [-CB {CB,CR,None}] [-mp MATCHPERCENT] [-temq TEMAPPINGQUALITY] [-exonmq EXONMAPPINGQUALITY] -i INPUT [INPUT ...] -idx INDEX [-o OUTPUT]

Get the expression of TE from single cell RNA sequencing data.

required arguments:
  -i INPUT [INPUT ...], --input INPUT [INPUT ...]
                        Please input the aligned file.
  -idx INDEX, --index INDEX
                        Give the path of index file.
  -o OUTPUT, --output OUTPUT
                        Prefix path of the output file.

optional arguments:
  -h, --help            show this help message and exit
  -CB {CB,CR,None}, --cellBarcode {CB,CR,None}
                        Location of barcodes. CB for cell ranger, CR for STAR. If they were seperated cell files from Smart-seq, select "None".
  -mp MATCHPERCENT, --matchPercent MATCHPERCENT
                        Matching threshold setting of reads.The higher the value, the more accurate. Default is 0.75.
  -temq TEMAPPINGQUALITY, --TEMappingQuality TEMAPPINGQUALITY
                        Threshold of mapping quality of TEs. Usually, TE reads have lower mapping quality.
  -exonmq EXONMAPPINGQUALITY, --exonMappingQuality EXONMAPPINGQUALITY
                        Threshold of mapping quality of reads on exon.

Example: TECT -i xxx.bam -idx human.index/index.idx -CB CB -o result
```

## Clust part

### input

* tematrix
  Use the count matrix generated in first step. It contains the TE information of each cell and total count of reads on exon.
* genematrix
  Use the gene matrix generated by cellranger.

### figures

* violin

  A violin plot of some of the computed quality measures:

  * the number of genes expressed in the count matrix

  * the total counts per cell

  * the percentage of counts in mitochondrial genes

* scatter

  Similar to the violin plot. Each point represents a cell.

You could adjust maxG(maxGene), maxC(maxCount) and mtp(mitochondria percentage) to filter the cells by the information in the violin and scatter plot.



* umap

  * umap_gene

    UMAP of gene expression

  * umap_te_cluster0

    UMAP of clusters that can be subdivided by te expression

  * umap_all_reproject

    Re-project the cluster divided by te onto the UMAP of gene

* rank_genes_groups

  * rank_genes_groups_leiden_before

    Marker genes of the different clust in the results of gene clustering

  * rank_genes_groups_leiden_te_cluster0

    The differentially expressed TE in the clusters which  secondary clustered

  * rank_genes_groups_te_leiden_after

    Marker genes of the different clust after add TE information

By the marker genes, you could distinguish cell type, and know which te caused the secondary clustering.

### Parameters

* gR

  Adjust the resolution of clustering by genes. The lower the parameter, the less the number of clusters given.

* teR

  Adjust the resolution of clustering by TEs. For each single cluster, subdivide accroding to the te expression.

### usage

```
TECT_clust -h                                                
usage: TECT [-h] [-o OUTPUT] [-maxG MAXGENE] [-maxC MAXCOUNT] [-mtp MTPERCENT]
            [-topG TOPGENE] [-gR GENERESOLUTION] [-teR TERESOLUTION] -te
            TEMATRIX -gene GENEMATRIX

Clustering the single-cell sequencing results.

required arguments:
  -te TEMATRIX, --TEMatrix TEMATRIX
                        Please input the TE count matrix file.
  -gene GENEMATRIX, --geneMatrix GENEMATRIX
                        the path of gene count matrix file.

optional arguments:
  -h, --help            show this help message and exit
  -o OUTPUT, --output OUTPUT
                        Prefix path of the output file.
  -maxG MAXGENE, --maxGene MAXGENE
                        Maximum number of genes expressed required for a cell
                        to pass filtering.
  -maxC MAXCOUNT, --maxCount MAXCOUNT
                        Maximum number of counts required for a cell to pass
                        filtering.
  -mtp MTPERCENT, --mtPercent MTPERCENT
                        Upper limit of mitochondrial content.
  -topG TOPGENE, --topGene TOPGENE
                        Number of highly-variable genes to keep.
  -gR GENERESOLUTION, --geneResolution GENERESOLUTION
                        Resolution of Leiden in gene clust.
  -teR TERESOLUTION, --teResolution TERESOLUTION
                        Resolution of Leiden in te clust.

Example: TECT -i xxx.bam -idx human.index/index.idx -CB CB -o result
```

## Example

1. Download the fastq file, process it with cellranger. Bam files sort by barcode tag.

   ```
   $ samtools sort -t CB -o bcsorted.bam xxx.bam
   ```

2. Download the te and exon annotation file by address above.

3. Run the TECT_build. Get the index file

   ```
   $ TECT_build -a te_annotation.out -g exon.gff3 -o human.index/
   ```

4. Run the TECT. Get the cell x TE matrix for the next step of analysis.

   ```
   $ TECT -i xxx.bam -idx human.index/index.idx -CB CB -blist xxx/outs/filtered_feature_bc_matrix/barcodes.tsv.gz -o result
   ```

5. Run the TECT_clust.

   ```
   $ TECT_clust -te count.txt -idx xxx/outs/filtered_feature_bc_matrix/
   ```

6. Downstream analysis on the clust result.

